### ISSUE TITLE
Feature: Handle properties that skip versions

### ISSUE THREAD

**Issue by theunrepentantgeek** [Azure/azure-service-operator#1776]
We need to design for the case where a particular property is removed in one version of a resource but reappears in a later version.

**TL;DR:** The crux of the problem is that we can currently end up with two different versions (shapes) of the same property serialised in a property bag on the same intermediate version.

To clarify, consider the following example. First we'll show the current behaviour, and then how that causes a problem. Finally, we'll discuss some constraints on any potential solution.

## Version 4

Consider a CRM system containing details of people. In v3 of the system, we capture each person's residental address, but in v4 we have dropped that property.

![image](https://user-images.githubusercontent.com/1272094/131927352-41115234-95d3-4f71-9adc-95b0d5cff244.png)

When we convert v3 -> v4, the `ResidentialAddress` property gets serialized into the `PropertyBag` on the v4 Person. Note that the bag contains a **v3** `Address` in a serialized form.

Conversion in the other direction (v4 -> v3) works fine, as we can deserialize the v3 `Address` from the `PropertyBag`.

## Version 5

In v5, the `ResidentialAddress` is reintroduced, but with a different shape. Instead of being the single field `Label`, it's now multiple fields.

![image](https://user-images.githubusercontent.com/1272094/131927399-d122a259-9c76-4737-aff2-d9d63e7945df.png)

When we convert v5 -> v4, again the `ResidentialAddress` property gets serialized into the `PropertyBag` on the v4 Person. Note that this time the bag contains a **v5** `Address` in a serialized form.

Conversion in the other direction (v4 -> v5) works fine, as we can deserialize the v5 `Address` from the property bag.

## The Problem

Observe how we can end up with two different kinds of v4 `Person`. We can end up with one where the property bag contains a **v3** `Address` or we can end up with one where the property bag contains a **v5** `Address`.

Here's where we run into problems.

![image](https://user-images.githubusercontent.com/1272094/131927447-30bf9eb5-493f-429b-8eab-e412173ec58d.png)

When we convert v3 -> v4 -> v5 we end up stuck part way through. 

After our first conversion step, we have a v4 `Person` that contains a v3 `Address` in the property bag. 

Conversion to the v5 `Person` will fail when we try to deserialize a **v5** `Address` from the **v3** version in the property bag. The two are not compatible, and the conversion will fail. This will cause the operator to fail.

### We can't see the future

We don't know what will happen with future versions - so we may already have users consuming ASO with v3 and/or v4 resources prior to the release of v5. 

However we choose to fix this problem, the introduction of a new resource version (v5 in our example) has to be backward compatible with resources that already exist prior to that release.

### Multi-generational skips

The example shown here has the new property returning after just one version; we need to handle the case where the property returns after a haiatis of several versions.

### Type changes

This example shows the shape of the complex type changing between versions. 

Consider how things would change if the `ResidentialAddress` on v3 `Person` was a simple string - instead of the property bag containing an object to deserialize, the v5 `Address` would find a simple string.

**Comment by matthchr** [comment]
No status change on this from above -- but we also haven't run into resources we've enabled that have this happen _yet_

**Comment by theunrepentantgeek** [comment]
Would be good to have our design thought out before the need arises.
